lib/bots/syncbot.ts

```diff
@@ -13,465 +13,345 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+// TODO: Test lots
+// TODO: Test in minikube
 
 import * as Promise from 'bluebird';
 import * as _ from 'lodash';
 import { ProcBot } from '../framework/procbot';
-import { Messenger } from '../services/messenger';
+import { MessengerService } from '../services/messenger';
 import {
-	DataHub,
-	FlowDefinition,
-	InterimContext, MessengerAction,
-	TransmitContext,
+	CreateThreadResponse, FlowDefinition, MessengerBasicMessageInformation,
+	MessengerEmitResponse, MessengerEvent, MessengerListenerMethod,
+	MessengerMessageAction, ThreadDefinition, TransmitInformation,
 } from '../services/messenger-types';
-import {
-	ServiceEvent,
-	ServiceListenerMethod,
-	ServiceRegistration,
-} from '../services/service-types';
-import { LogLevel } from '../utils/logger';
+import { createDataHub } from '../services/messenger/datahubs/datahub';
+import { Logger, LogLevel } from '../utils/logger';
 
+/**
+ * A bot that mirrors threads across services.
+ */
 export class SyncBot extends ProcBot {
 	/**
-	 * Returns an array of the tokens in a provided context
-	 * @param event  Context to data mine for tokens
-	 */
-	private static extractTokens(event: InterimContext) {
-		const secrets = [];
-		if (event.toIds.token) {
-			secrets.push(event.toIds.token);
-		}
-		if (event.sourceIds.token) {
-			secrets.push(event.sourceIds.token);
-		}
-		return secrets;
-	}
-
-	// These objects store built objects, typed and indexed, to help minimise object rebuilding
-	private messengers = new Map<string, Messenger>();
-	private hub: DataHub;
-
-	/**
-	 * Creates a SyncBot using SYNCBOT_MAPPINGS and SYNCBOT_HUB_SERVICE from the environment.
-	 * @param name  Identifier for this bot, defaults to SyncBot.
-	 */
-	constructor(name = 'SyncBot') {
-		super(name);
-		// Register each edge in the mappings array bidirectionally
-		const mappings: FlowDefinition[][] = JSON.parse(process.env.SYNCBOT_MAPPINGS);
-		for (const mapping of mappings) {
-			let priorFlow = null;
-			for (const focusFlow of mapping) {
-				if (priorFlow) {
-					this.register(priorFlow, focusFlow);
-					this.register(focusFlow, priorFlow);
-				}
-				priorFlow = focusFlow;
-			}
-		}
-	}
-
-	/**
-	 * Awaken services and register the event processors.
-	 * @param from  Definition of a flow to listen to.
-	 * @param to    Definition of a flow to emit to.
+	 * Provide a method that:
+	 *  * encloses details available in advance.
```

Comment:
This is because the clause I was writing:
* Felt like a continuation of the introductory sentence.
* Did not feel like a complete sentence in its own right.

I will:
* Fix this in the code.
* Improve my grammar.